ocfweb/docs/docs/services/hpc.md

```diff
@@ -0,0 +1,115 @@
+
+```
+virtualenv --system-site-packages -p python3 venv
+. venv/bin/activate
+```
+
+This will allow you to *pip install* any Python packages that the OCF does not already have for your program.
+
+### Singularity
+
+For those who need access to non-Python dependencies or have already integrated their program into Docker, the second option is to use Singularity containers. [Singularity][singularity] is a containerization platform developed at Lawrence Berkeley National Laboratory that is designed specifically for HPC environments. To read more about the benefits of Singularity you can look [here][singularity_article]. We suggest a particular workflow, which will help simplify deploying your program on our infrastructure.
+
+#### Installing
+
+If you are running linux on your system, you can install Singularity from the official apt repos:
+
+```
+sudo apt install singularity-container
+```
+
+If you are running Mac you can look [here][mac_install], or Windows [here][win_install].
+
+#### Building Your Container
+
+```
+singularity build --sandbox ./my_container docker://ubuntu
+```
+This will create a Singularity container named `my_container`. If you are still testing out your container on your system, we suggest you use the `--sandbox` option. This will allow you to install new packages in your container as you need them without needing sudo privileges. The `docker://ubuntu` option notifies Singularity to bootstrap the container from the official Ubuntu docker container on [Docker Hub][docker_hub]. There is also a [Singularity Hub][singularity_hub], from which you can directly pull Singularity images in a similar fashon. We also have some pre-built containers that you may use to avoid having to build your own. They are currently located at `/home/containers`.
+
+#### Using Your Container
+
+```
+singularity shell my_container
+```
+The above command will allow you to shell into your container. By default your home directory in the container is linked to your real home directory outside of the container environment, which helps you avoid having to transfer files in and out of the container.
+
+```
+singularity exec --nv my_container ./my_executable.sh
+```
+This command will open your container and run the `my_executable.sh` script in the container environment. The `--nv` option allows the container to interface with the GPU. This command is useful when using `srun` so you can run your program in a single command.
+
+#### Moving to HPC Infrastructure
+
+If you were using a sandboxed container for testing, we suggest you convert it to a Singularity image file. This is because images are more portable and easier to interact with than sandboxed containers. You can make this conversion using the following command:
+```
+sudo singularity build my_image.simg ./my_sandboxed_container
```

Comment:
One thing that's not quite clear here is if the user is intended to build the sandboxed container on the hpc host or on his/her desktop. This is because earlier you provided instructions for installing `singularity-container` on Windows + Mac. I imagine both work, so it might be worth mentioning that they need to `scp` or `stfp` the files to the hpc host if they built the sandboxed container on their desktop.

Reply:
Perhaps we should encourage users to just log into segfault to develop, rather than setting something up on their personal machine, debugging it, transfering it to our machine, then debugging it again?

Reply:
Yeah this is something we have not been totally sure on. @redplanks suggested developing on personal machines because I believe he saw some singularity developers recommended it. I could go either way, the only issue I see with developing on our infra is that they may use up too many resources they don't need (because their code is buggy and not working correctly, etc.) when there are others who have their stuff together and need to get things run in a timely manner.

Reply:
I would say have them develop it on the HPC host for now. While it's nice to provide people with options, I think it's already enough new information such that HPC users will feel overwhelmed. You could maybe add the Mac and Windows install instructions as an addendum, but recommend doing it on the host. 

If it does become an issue and HPC is hugely popular we can always revise the documentation. Since HPC has its own `nfs` we could even set aside another machine for building `singularity` images.